chirp/drivers/baofeng_uv17Pro.py

```diff
@@ -0,0 +1,1039 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+
+from chirp.drivers import baofeng_common
+from chirp import chirp_common, directory, memmap
+from chirp import bitwise
+from chirp.settings import RadioSetting, \
+    RadioSettingValueBoolean, RadioSettingValueList, \
+    RadioSettingValueString, \
+    RadioSettings, RadioSettingGroup
+import struct
+from chirp import errors, util
+
+LOG = logging.getLogger(__name__)
+
+# Baofeng UV-17L magic string
+MSTRING_UV17L = b"PROGRAMBFNORMALU"
+MSTRING_UV17PROGPS = b"PROGRAMCOLORPROU"
+
+DTMF_CHARS = "0123456789 *#ABCD"
+STEPS = [2.5, 5.0, 6.25, 10.0, 12.5, 20.0, 25.0, 50.0]
+
+LIST_AB = ["A", "B"]
+LIST_ALMOD = ["Site", "Tone", "Code"]
+LIST_BANDWIDTH = ["Wide", "Narrow"]
+LIST_COLOR = ["Off", "Blue", "Orange", "Purple"]
+LIST_DTMFSPEED = ["%s ms" % x for x in [50, 100, 200, 300, 500]]
+LIST_HANGUPTIME = ["%s s" % x for x in [3, 4, 5, 6, 7, 8, 9, 10]]
+LIST_SIDE_TONE = ["Off", "Side Key", "ID", "Side + ID"]
+LIST_DTMFST = ["Off", "DT-ST", "ANI-ST", "DT+ANI"]
+LIST_MODE = ["Name", "Frequency", "Channel Number"]
+LIST_OFF1TO5 = ["Off"] + list("12345")
+LIST_OFF1TO9 = LIST_OFF1TO5 + list("6789")
+LIST_OFF1TO10 = LIST_OFF1TO9 + ["10"]
+LIST_OFFAB = ["Off"] + LIST_AB
+LIST_RESUME = ["TO", "CO", "SE"]
+LIST_PONMSG = ["Full", "Message"]
+LIST_PTTID = ["Off", "BOT", "EOT", "Both"]
+LIST_SCODE = ["%s" % x for x in range(1, 21)]
+LIST_RPSTE = ["Off"] + ["%s" % x for x in range(1, 11)]
+LIST_SAVE = ["Off", "1:1", "1:2", "1:3", "1:4"]
+LIST_SHIFTD = ["Off", "+", "-"]
+LIST_STEDELAY = ["Off"] + ["%s ms" % x for x in range(100, 1100, 100)]
+LIST_STEP = [str(x) for x in STEPS]
+LIST_TIMEOUT = ["Off"] + ["%s sec" % x for x in range(15, 195, 15)]
+LIST_TIMEOUT_ALARM = ["Off"] + ["%s sec" % x for x in range(1, 11)]
+LIST_PILOT_TONE = ["1000 Hz", "1450 Hz", "1750 Hz", "2100 Hz"]
+LIST_TXPOWER = ["High", "Low"]
+LIST_VOICE = ["Off", "English", "Chinese"]
+LIST_WORKMODE = ["Frequency", "Channel"]
+LIST_POWERON_DISPLAY_TYPE = ["LOGO", "BATT voltage"]
+LIST_BEEP = ["Off", "Beep", "Voice", "Both"]
+LIST_LANGUAGE = ["English", "Chinese"]
+LIST_SCANMODE = ["Time", "Carrier", "Search"]
+LIST_ALARMMODE = ["Local", "Send Tone", "Send Code"]
+LIST_MENU_QUIT_TIME = ["%s sec" % x for x in range(5, 55, 5)] + ["60 sec"]
+LIST_BACKLIGHT_TIMER = ["Always On"] + ["%s sec" % x for x in range(5, 25, 5)]
+LIST_ID_DELAY = ["%s ms" % x for x in range(100, 3100, 100)]
+LIST_QT_SAVEMODE = ["Both", "RX", "TX"]
+LIST_SKEY2_SHORT = ["FM", "Scan", "Search", "Vox"]
+LIST_RPT_TAIL_CLEAR = ["%s ms" % x for x in range(0, 1100, 100)]
+LIST_VOX_DELAY_TIME = ["%s ms" % x for x in range(500, 2100, 100)]
+LIST_VOX_LEVEL = ["Off"] + ["%s" % x for x in range(1, 10, 1)]
+LIST_VOX_LEVEL_ALT = ["%s" % x for x in range(1, 10, 1)]
+LIST_GPS_MODE = ["GPS", "Beidou", "GPS + Beidou"]
+LIST_GPS_TIMEZONE = ["%s" % x for x in range(-12, 13, 1)]
+CHARSET_GB2312 = chirp_common.CHARSET_ASCII
```

Comment:
Are you just doing this to limit the charset for the setting string? If so, I think it would be better (i.e. more conventional) to just let it be a string (which is unicode) and then encode it to gb2312 when we store it. If they use a character that isn't supported, we can let the UnicodeEncodeError bubble up to the user to alert them.

Reply:
I am defining the GB2312 charset here and the following lines. Later I use it for displaying the Chinese characters in the bank names. It was a hell of a job to figure out how to make this work. I have no clue how to make this better.

Reply:
What's the actual width of the field of potential characters? Is it 16 or more? If it's 16, then use a `u8` in the memory format, then we can collect the bytes and decode them like this I think:
```
bank_raw = bytes([int(x) for x in _mem.banknames[0].name])
bank_decoded = charbytes.decode('gb2312')
```
I think that should be fully clean and the reverse will work for encoding.

Reply:
That's basically what I am doing in _filtername.

Reply:
I thought you said it doesn't decode properly when there are multi-byte characters? You're not passing the full list of bytes to the decode when you get to a `0xFF`, which may or may not be right. I'm not sure what the raw encoding of `gb2312` looks like. Either way, treating those as a `char` and doing the `chr(ord(c))` stuff is mixing the interpretations of the contents back and forth between raw bytes and unicode strings in python. Treating them as ints, then raw bytes, and then performing the decode to unicode as a whole is the proper way.